nixos/modules/services/web-apps/nextcloud-backup.nix

```diff
@@ -0,0 +1,196 @@
+      {
+        # The sqlite database is automatically backed up since it's contained in the datadir
+        backupFileName = "data/${cfg.config.dbname}.db";
+        backupCommand = "";
+        restoreCommand = "";
+      }
+    else if cfg.config.dbtype == "pgsql" then
+      let
+        backupFileName = "database-pgsql.bak";
+        pg_dump = "${config.services.postgresql.package}/bin/pg_dump";
+        pg_restore = "${config.services.postgresql.package}/bin/pg_restore";
+        setPassword =
+          if cfg.config.dbpassFile != null then
+            ''
+              export PGPASSWORD="$(cat ${cfg.config.dbpassFile})"
+            ''
+          else
+            "";
+        commonFlags = "-h ${cfg.config.dbhost} -U ${cfg.config.dbuser} --no-password";
+      in
+      {
+        inherit backupFileName;
+        backupCommand = ''
+          ${setPassword}
+          ${pg_dump} ${commonFlags} -F custom -f $1/nextcloud-backup/${backupFileName} ${cfg.config.dbname}
+        '';
+        restoreCommand = ''
+          ${setPassword}
+          sudo -E -u nextcloud ${pg_restore} ${commonFlags} --clean --if-exists -d ${cfg.config.dbname} $1/nextcloud-backup/${backupFileName}
+        '';
+      }
+    else if cfg.config.dbtype == "mysql" then
+      let
+        backupFileName = "database-mysql.bak";
+        mysqldump = "${config.services.mysql.package}/bin/mysqldump";
+        mysql = "${config.services.mysql.package}/bin/mysql";
+        # MySQL doesn't accept the socket path as a hostname
+        hostFlag = if cfg.database.createLocally then "-h localhost" else "-h ${cfg.config.dbhost}";
+        passwordFlag =
+          if cfg.config.dbpassFile != null then
+            ''
+              -p "$(cat ${cfg.config.dbpassFile})"
+            ''
+          else
+            "";
+        commonFlags = hostFlag + " -u ${cfg.config.dbuser} " + passwordFlag;
+      in
+      {
+        inherit backupFileName;
+        backupCommand = ''
+          ${mysqldump} ${commonFlags} --add-drop-table --single-transaction ${cfg.config.dbname} > $1/nextcloud-backup/${backupFileName}
+        '';
+        restoreCommand = ''
+          sudo -u nextcloud ${mysql} ${commonFlags} ${cfg.config.dbname} < $1/nextcloud-backup/${backupFileName}
+        '';
+      }
+    else
+      throw "Unknown nextcloud database type ${cfg.config.dbtype}";
+  checkBackup = ''
+    for f in \
+      "$1/nextcloud-backup/config.php" \
+      "$1/nextcloud-backup/${dbBackupRestore.backupFileName}" \
+      "$1/nextcloud-backup/store-apps" \
+      "$1/nextcloud-backup/data"
+    do
+      if [ ! -e "$f" ] ;then
+        echo "Did not find file $f."
+        echo "It looks like the backup is incomplete, aborting."
+        exit 1
+      fi
+    done
+  '';
+  backupScript = pkgs.writeShellScriptBin "nextcloud-backup" ''
+    set -Eeuo pipefail
+    if [ "$#" != "1" ] || [ "$1" = "-h" ] || [ "$1" = "--help" ] ;then
+      echo "Usage: nextcloud-backup TARGET_DIR"
```

Comment:
It feels like this script is somewhere between a service and a package. It exists as a service, but then some options, like the `TARGET_DIR` aren't settable using nixos but need to be given on the command line. I think you should embrace the service part more, make all inputs configurable as nixos options, and maybe allow overrides on the command line. This would make it easier to use the script inside other modules, and might allow us for example to add our own timed service for backups at a later point.